scripts/usb_scan.py

```diff
@@ -0,0 +1,660 @@
+            allow_interface = False
+            for r in self.rule_list:
+                if self.match_device_interface(r, device, i):
+                    if r[self._ALLOW]:
+                        log_wrapper("allow " + i.get_node())
+                        allow_interface = True
+                        break
+                    else:
+                        log_wrapper("deny " + i.get_node())
+                        return False
+            if not allow_interface:
+                log_wrapper("deny " + i.get_node() + ", no matching rule")
+                return False
+
+        log_wrapper("allow " + device.get_node())
+        return True
+
+
+def parse_args():
+    parser = argparse.ArgumentParser(
+        description="scanner to get USB devices info")
+    parser.add_argument("-d", "--diagnostic", dest="diagnostic",
+                        action="store_true",
+                        help="enable diagnostic mode")
+    return parser.parse_args()
+
+
+def to_pusb(device):
+    """ convert UsbDevice to pusb dict
+
+    Example pusb dict:
+    [
+      {
+        "product-desc": "",
+        "product-id": "383c",
+        "description": "Lenovo_H8RU2TGE",
+        "vendor-desc": "Lenovo",
+        "version": "2.00",
+        "vendor-id": "17ef",
+        "path": "1-2",
+        "serial": "H8RU2TGE"
+      },
+      {
+        "product-desc":
+           "TransMemory-Mini / Kingston DataTraveler 2.0 Stick (2GB)",
+        "product-id": "6544",
+        "description":"Toshiba Corp._TransMemory-Mini /
+            Kingston DataTraveler 2.0 Stick (2GB)_9875B1B3F9F4CD301917964B",
+        "vendor-desc": "Toshiba Corp.",
+        "version": "2.00",
+        "vendor-id": "0930",
+        "path": "4-1.2",
+        "serial": "9875B1B3F9F4CD301917964B"
+      }
+    ]
+    :param device:(UsbDevice) the device to convert
+    :return:(dict) the key value pairs for pusb
+    """
+    pusb = {}
+
+    pusb["path"] = device.get_node()
+    # strip the leading space of "version"
+    pusb["version"] = device[UsbDevice._VERSION].strip()
+    pusb["vendor-id"] = device[UsbDevice._ID_VENDOR]
+    pusb["product-id"] = device[UsbDevice._ID_PRODUCT]
+    pusb["vendor-desc"] = device[UsbDevice._DESC_VENDOR]
+    pusb["product-desc"] = device[UsbDevice._DESC_PRODUCT]
+    # strip serial due to CA-268761
+    pusb["serial"] = device[UsbDevice._SERIAL].strip()
+
+    pusb["description"] = pusb["vendor-desc"]
+    if pusb["product-desc"]:
+        pusb["description"] += "_" + pusb["product-desc"]
+    if pusb["serial"]:
+        pusb["description"] += "_" + pusb["serial"]
+
+    return pusb
+
+
+if __name__ == "__main__":
+    args = parse_args()
+    if args.diagnostic:
+        __DIAGNOSTIC = True
+
+    # get usb info
+    devices, interfaces = get_usb_info()
+
+    # debug info
+    log_list_wrapper(devices)
+    log_list_wrapper(interfaces)
+
+    # match interface to device
+    for i in interfaces:
+        for d in devices:
```

Comment:
I think this is more readable (and faster although that's not really a problem here) as

```python
for device, interface in itertools.product(devices, interfaces):
  if nterfacei.is_child_of(device):
    device.add_interface(interface)
```

This is only a style preference, feel free not to do this change.

Reply:
thanks. please allow me to skip this one for now.

Reply:
Sure, this is not important